subsys/net/lib/zzhc/zzhc.c

```diff
@@ -451,6 +454,80 @@ static int check_iccid(struct zzhc *ctx)
 	return rc;
 }
 
+static int check_and_set_dns(struct zzhc *ctx)
+{
+	int err;
+	uint8_t dns_compare[32];
+	struct nrf_in_addr pri_dns;
+	struct nrf_in_addr sec_dns;
+	struct nrf_addrinfo *info;
+	struct nrf_addrinfo hints = {
+		.ai_family = NRF_AF_INET,
+		.ai_socktype = NRF_SOCK_STREAM,
+		.ai_protocol = NRF_IPPROTO_TCP,
+	};
+
+	err = nrf_modem_at_cmd(ctx->at_resp, AT_RESPONSE_LEN,
+			       "AT+CGCONTRDP=0");
+
+	if (err) {
+		return -EIO;
+	}
+
+	snprintk(dns_compare, sizeof(dns_compare),
+		 "\"%s\",\"%s\"",
+		 PRIMARY_DNS,
+		 SECONDARY_DNS);
+
+	if (strstr(ctx->at_resp, dns_compare)) {
+		return 0;
+	}
+
+	err = nrf_inet_pton(NRF_AF_INET, PRIMARY_DNS, &pri_dns);
+
+	if (err != 1) {
+		LOG_DBG("nrf_inet_pton failed. errno: %d", errno);
+		return -EIO;
+	}
+
+	/* Manually set DNS */
+	err = nrf_setdnsaddr(NRF_AF_INET, &pri_dns, sizeof(pri_dns));
+
+	if (err) {
+		LOG_DBG("nrf_setdnsaddr failed. errno: %d", errno);
+		return -EIO;
+	}
+
+	/* Hostname resolution */
+	err = nrf_getaddrinfo(hostname, NULL, &hints, &info);
+	if (err) {
+		LOG_DBG("Can't resolve hostname %s. err: %d. Trying secondary backup DNS.",
+			log_strdup(hostname), err);
+	}
+
```

Comment:
Logic below this should be included in if (err)

Reply:
@junqingzou Thanks, forgot to add the `return 0` in case of positive outcome.